### DISCUSSION TITLE
Are the Spark.NET 3.5.1 works in Polyglot notebook for UDF examples?

### DISCUSSION THREAD

**Discussion by GeorgeS2019** [dotnet/spark#1179] [General]
https://github.com/dotnet/spark/issues/1199
![image](https://github.com/user-attachments/assets/cfca168f-792d-482f-869e-97ce74723bb5)

@grazy27
[4. User-Defined Functions](https://learning.oreilly.com/library/view/introducing-net-for/9781484269923/html/502681_1_En_4_Chapter.xhtml)

Previously in older version, it is not possible to work in the UDF notebook. Not sure if the latest 3.5.1 overcomes this limitation

https://github.com/Apress/introducing-.net-for-apache-spark
![image](https://github.com/user-attachments/assets/3e851969-393c-4ea0-a5cd-1b5b41837955)

**Comment by grazy27** [comment]
@GeorgeS2019 It does not. 

It tries to get entire declaring type, which is anonymous assembly that only exists within interactive session context. 
And fails with an exception, as result:
![image](https://github.com/user-attachments/assets/3230668e-3078-4725-a96c-fd045cdbcb22)

Code expects declaring assembly for UDF to be accessible: 
![image](https://github.com/user-attachments/assets/91dd7228-6bee-413f-939a-e9197e240769)

Notebooks without UDFs and broadcasts work fine.

**Reply by GeorgeS2019** [reply]
Do U have recommendation to the polyglot team what changes they need to make to accommodate spark.net UDF to work?  

I am going to invite them to address your findings 

Good job

**Reply by grazy27** [reply]
Certainly, there are ways to achieve this.
In the context of Dotnet.Spark, we can determine whether the current code is running in an interactive environment and manually include library bytes in the command payload (or alternatively, save the library as a file and load it on the worker side).

PoC: 
![image](https://github.com/user-attachments/assets/b78ae897-9316-4af0-af0c-692c20db295b)
For this PoC to work, we also need to hardcode executing assembly as 1.dll 
![image](https://github.com/user-attachments/assets/bf98fc90-e867-4838-89dc-360f1d696001)

Polyglot team could do something similar on their side as well, but it'd have to be strongly names and uploaded to GAC, or explicitly target '%DOTNET_WORKER_DIR%'. 
This would only work locally, and other dotnet.Interactive wrappers would need the same fix. 

This issue falls primarily within Dotnet.Spark's scope of responsibility. This issue should be taken into account when addressing migration from BinarySerializer to somewhere else, since the responsible code will have to be rewritten.

**Comment by GeorgeS2019** [comment]
https://github.com/dotnet/interactive/issues/3797

**Reply by wudanzy** [reply]
I guess pyspark did save all bytecode? We can do the same in spark.NET.

**Comment by grazy27** [comment]
I’ve made some progress in understanding the interaction, though it’s still not entirely clear. Here’s what I’ve gathered so far:

Dotnet.Spark contains a separate solution for this already, `Microsoft.Spark.Extensions.DotNet.Interactive`. It registers into .net interactive pipeline and adds all executables via 'spark.addFile' functionality.  Does pretty much the same I did in my PoC, but also handles more special cases and uses native spark function for it. 
That's clever, we don't have to serialize IL code and send it to spark.worker.
On a worker side, there's dedicated code for handling this as well. 

The only problem i faced is that it doesn't work, it relies on Dotnet.Interactive API `Utility.DisposableDirectory` API, using which throws `MethodAccessException`. 
![image](https://github.com/user-attachments/assets/0e6cf16d-841d-45bb-b346-bc2f508888e3)
According to [this issue in dotnet interactive](https://github.com/dotnet/interactive/issues/3136), is not ready for 1.0 stable release. 

Thanks @GeorgeS2019 for creating issue in interactive, let's wait and see if they can point us with this exception.

**Comment by jonsequitur** [comment]
The constructor is private. But also this class will likely be made internal as part of API stabilization. The .NET Interactive API shouldn't really be responsible for providing this kind of utility code to consumers. I'd recommend cloning the functionality as needed.

https://github.com/dotnet/interactive/blob/6ae92454e21a8793cb289e6d70508bd5a4e73418/src/Microsoft.DotNet.Interactive/Utility/DisposableDirectory.cs

**Comment by grazy27** [comment]
Hello @jonsequitur,

I've investigated it in more detail, and have a few findings and questions. 
Would you mind reviewing the proposed solution for `2` and providing advice for `3`? Thank you in advance!

## 1. DisposableDirectory 
The DisposableDirectory issue is resolved by using kernel.RegisterForDisposal for manual directory cleanup, which aligns with the current API.

## 2. Collection of referenced nugets and their dependencies
For NuGet dependency collection, ISupportNuget has been removed, breaking the previous method of getting ResolvedPackageReferences ([SupportNugetWrapper](https://github.com/dotnet/spark/blob/main/src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/SupportNugetWrapper.cs), that is utilized in [PackageResolver](https://github.com/dotnet/spark/blob/main/src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/PackageResolver.cs)).

Workaround: I am now subscribing to the PackageAdded event to collect ResolvedPackageReference objects, which we need for probing paths and package metadata.
Limitation: This approach requires our extension to be loaded before any other NuGet packages to guarantee a complete collection.
**Question:** Is this event subscription the recommended pattern, or is there a more robust method to get all resolved packages, including those potentially loaded before our extension?


## 3. The middleware pipeline behavior for submissions appears to have changed, there's a timing issue
 In [AssemblyExtension](https://github.com/dotnet/spark/blob/f28580410a643c27eee74385f774057448640e46/src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/AssemblyKernelExtension.cs#L51) there's a middleware registration, that first propagates down the pipeline `await next(....)`, and only then accesses ScriptContext and attaches a new compiled `Submission-x.dll` file to spark. 
 
 I'm not sure how it worked before, but now it does not - by the time we can access dll, submission have already failed as there was no dll.
 
If we add the DLL before `await next(...)`, kernel.ScriptState is stale and provides the DLL from the previous submission (n-1), causing a version mismatch on the worker.

 If I first attach dll, and then propagate the pipeline - ScriptContext contains `Submission(X-1).dll` - Previous cell value, and this one doesn't work as well, as dotnet.spark worker fails with version mismatch.

**Question**: What is the correct pattern to inject logic after compilation but before execution to prepare the Spark Context with the current submission's assembly?

**Comment by jonsequitur** [comment]
> Question: Is this event subscription the recommended pattern, or is there a more robust method to get all resolved packages, including those potentially loaded before our extension?

You can see the loaded metadata references like this:

![image](https://github.com/user-attachments/assets/51f7bf64-165c-4be6-ab52-10e064bcc82a)

**Reply by grazy27** [reply]
This approach will only work for basic scenarios, as more complex nuget libraries may contain references to native C++ dlls, or may require additional non-binary files, like `.xml\.json\.config`, and they will not be copied in such case

[Current implementation](https://github.com/dotnet/spark/blob/f28580410a643c27eee74385f774057448640e46/src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/PackageResolver.cs#L46) includes code for extracting those references from nuget metadata and sending them to remote worker as well, but this additional is only available in `nuget package` object

**Reply by jonsequitur** [reply]
You probably need to get at the [`PackageRestoreContext`](https://github.com/dotnet/interactive/blob/1c35640c1803e586fede30ad4b4738326d83361c/src/Microsoft.DotNet.Interactive.PackageManagement/PackageRestoreContext.cs#L18) if you want the rest of that information.  It looks a bit like the .NET for Spark implementation you pointed to was based on this code, several refactors ago. 

We might need to expose an API to make the `PackageRestoreContext` accessible. While the type is public, the instance associated with a given kernel is scoped in such a way that getting a reference to it would be difficult: 

https://github.com/dotnet/interactive/blob/c125401b58a9069d4b186600365aedba7e3d04c0/src/Microsoft.DotNet.Interactive.PackageManagement/KernelExtensions.cs#L16

**Reply by grazy27** [reply]
Aha, then i can try to get it from internal collection of disposables from Kernel using reflection magic

This will win you some time to add a way to get it properly from the context, and we can migrate to normal approach next release

**Comment by jonsequitur** [comment]
> 3. The middleware pipeline behavior for submissions appears to have changed, there's a timing issue

I would need more context here. Do you have a stack trace? Is the DLL that's missing the one that's supposed to be emitted from the `CSharpKernel`'s compilation?

I'm not sure where the timing issue could be introduced. There's no parallelism in the scheduler.

> Question: What is the correct pattern to inject logic after compilation but before execution to prepare the Spark Context with the current submission's assembly?

This code looks reasonable. Another approach that you could try would be to subscribe to `CommandSucceeded` events rather than using middleware.

**Reply by grazy27** [reply]
Thanks for your answers!
You're right, dll is emitted from compilation -> added to spark context -> copied to spark temp folder -> used by spark worker to run the `func` object used as UDF

If there were no changes in threading or parallelism, I can't understand how(If?) it worked before
Here's [current impl](https://github.com/dotnet/spark/blob/f28580410a643c27eee74385f774057448640e46/src/csharp/Extensions/Microsoft.Spark.Extensions.DotNet.Interactive/AssemblyKernelExtension.cs#L51), supposedly it worked before
 

The problem I'm facing is that a new assembly needs to be added before command execution, but it's only available in compilation afterwards 

![image](https://github.com/user-attachments/assets/d17cef81-305a-4582-a1a3-f28e1360cc11)


Is there any way to get new `Compilation` object before it get's executed?

**Reply by jonsequitur** [reply]
> Is there any way to get new Compilation object before it get's executed?

No, compilation and execution happen together within Roslyn scripting.

**Reply by jonsequitur** [reply]
Can you provide an example of a code cell containing a UDF?

**Reply by grazy27** [reply]
Sure, 
here's simple example for testing: 
``` CSharp
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;

using static Microsoft.Spark.Sql.Functions;

var spark = SparkSession.Builder().GetOrCreate();

string IntToStr(int number)
{
    return $"The number is {number}";
}
Func<Column, Column> udfIntToStr = Udf<int, string>(IntToStr);

var df = spark.Sql("Select ID from range(1000)");
df.Select(udfIntToStr(df.Col("ID"))).Show();
```
Worker needs to access definition of the `IntToStr` to run it, there is a [wiki page in dotnet.spark](https://github.com/dotnet/spark/blob/main/docs/understanding-microsoft.spark.md) to help with visualizing how it works 



<br><br>
My prod-like use case will also rely on nugets, and work with batches of data: 

``` CSharp
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Ihor.ComplexLibraryThatDoesSomeMagic;

using static Microsoft.Spark.Sql.Functions;

var spark = SparkSession.Builder().GetOrCreate();

// This entire func is serialized and executed on a remote machine
RecordBatch CountCharacters(RecordBatch records)
{
   var newRb = PreparerecordBatch(rb);
   var lib = new ComplexLibraryThatDoesSomeMagic(configs);
   return lib.DoSomeMagic(newRb);
}

var df = spark.Load.Parquet("local path");

df.GroupBy("Year").Apply(
    new StructType(new[]
    {
        new StructField("Year", new IntegerType()),
        new StructField("SalesData", new ArrayType(new IntegerType())),
        ....... more
    }),
    calculateSomeMetricsPerYear())
.Show();
```

**Reply by grazy27** [reply]
Is it possible to compile it two times: First before propagating the pipeline, without triggering actual run, to use that compilation for emitting dll
And second with `next(cmd, ctx)`?

**Reply by jonsequitur** [reply]
It's not. Compilation and execution happen in the same method call at the C# Script level.

**Reply by grazy27** [reply]
Got it, it's a dealbreaker from dotnet perspective.
Is there any chance to change smth in dotnet.interactive to allow that to work? 
Mb events like 'OnBeforeExecutionStarted\OnAfterCompiled'?

How would you recommend to approach solving this?

**Reply by jonsequitur** [reply]
This worked before without this ability so it should still be possible.

Trying to do this in middleware might be part of the problem, though it's not clear to me. An occasional gotcha to be aware of here is that if you submit code to the `CSharpKernel` during the execution of code within the same kernel, you end up with two different assemblies, one of which (generated by the inner submission) is abandoned when the outer submission completes. This is because compilations are immutable and only one (the last to complete) is left available for subsequent submissions to build on.

So rather than trying to do this in middleware, I would recommend subscribing to all `CommandSucceeded` events at the kernel level and after each one, capture the assembly.

**Reply by grazy27** [reply]
Thanks for your help! 
Two different but identical ones sounds Ok, as long as they're identical
If there's a conflicting assembly executed, and current dismissed - that should also do, as they'll have different signatures

I'll have a look at events to subscribe, and will try to understand how it worked before, as according to all logic it shouldn't have - assembly is needed before command execution, there's little help from it afterwards

The only option that comes to my mind - if it only worked with deferred execution:
Cell 1
``` csharp
var df = spark.Read().Parquet........UDF.....
``` 
Cell 2
``` csharp
df.collect()
Or df.print
``` 
Then assembly will be generated and added before execution, as execution happens in a different command

If that's the case, I'll migrate to `CommandSucceeded` to make it more transparent and stable


@GeorgeS2019  do you have any info on how it used to work previously?

**Reply by jonsequitur** [reply]
> Two different but identical ones sounds Ok, as long as they're identical
> If there's a conflicting assembly executed, and current dismissed - that should also do, as they'll have different signatures

They're not identical. They will typically be completely different. Let me give an example.

![image](https://github.com/user-attachments/assets/95029980-e69c-45a5-a1aa-d8fae300f584)

In this example, the outer submission is all of the code in the cell. The inner submission is the code on lines 8 and 9 that are being sent to the same `CSharpKernel` instance.

The complete sequence of events here looks like this:

1. This code (let's call it the "outer submission") is sent to the `CSharpKernel`:

    ```csharp
    using Microsoft.DotNet.Interactive;
    using Microsoft.DotNet.Interactive.Commands;
    using Microsoft.DotNet.Interactive.Events;

    var csharpKernel = Kernel.Root.FindKernelByName("csharp");

    var innerResult = await csharpKernel.SendAsync(new SubmitCode("""
        var innerVariable = "Inner variable"; 
        Console.WriteLine(innerVariable);
        """));

    innerResult.Events
    ```

2. This code is compiled and executed (a single step within C# Scripting).

3. The executing code sends another command to the `CSharpKernel` which compiles and executes the following code (let's call it the "inner submission"):

    ```csharp
    var innerVariable = "Inner variable"; 
    Console.WriteLine(innerVariable);
    ```

4. The `CSharpKernel` completes the handling of the `SubmitCode` command carrying the _**inner**_ submission and stores the resulting assembly as the most-recently-executed assembly in the chain.

5. The `CSharpKernel` completes the handling of the `SubmitCode` command carrying the _**outer**_ submission and stores the resulting assembly as the most-recently-executed assembly in the chain. This effectively discards the inner submission's assembly. 

As you can see in the variable view at the bottom of the screenshot, the variable `innerVariable` is not present. It was only declared in the inner submission's assembly, which is gone by the time the outer submission is done being executed.

So the two different assemblies represent a fork. They will typically contain no overlapping code.

**Reply by grazy27** [reply]
Interesting, thanks for sharing this combination!
Dotnet.spark worker only needs an entire assembly,  so if we subscribe to `CommandSucceeded`, in your case we'll get 2 different invocations and 2 different assemblies, right? 
And one of them is discarded afterwards and global state is not updated
In such case, we can anyway attach both of them to the spark context, and discarded one will just never be referenced from worker, and cleansed after run anyway - I'll make a small test to doublecheck that


Developers will do 5 hours of debugging to save 10 min of reading the documentation 🙂
I found [an official guide for UDFs in interactive .net.spark,](https://learn.microsoft.com/en-us/previous-versions/dotnet/spark/how-to-guides/dotnet-interactive-udf-issue#call-a-udf-on-a-dataframe) and they specifically mentioned to run UDF creation in one cell, and its invocation in another, so we can proceed with implementation
